crates/aptos-logger/src/speculative_log.rs

```diff
@@ -0,0 +1,284 @@
+// Copyright (c) Aptos
+// SPDX-License-Identifier: Apache-2.0
+
+use crate as aptos_logger;
+/// Infrastucture for logging speculatively, i.e. ability to clear prior logs.
+use crate::{debug, error, info, trace, warn, Level};
+use aptos_crypto::HashValue;
+use aptos_infallible::Mutex;
+use aptos_log_derive::Schema;
+use aptos_state_view::StateViewId;
+use aptos_types::transaction::Version;
+use arc_swap::ArcSwap;
+use crossbeam::utils::CachePadded;
+use once_cell::sync::Lazy;
+use rayon::prelude::*;
+use serde::Serialize;
+use std::sync::{
+    atomic::{AtomicUsize, Ordering},
+    Arc,
+};
+
+#[derive(Default)]
+struct BufferedLog {
+    entries: Vec<(Level, AdapterLogSchema, String)>, // Log level & the message string
+    critical_err_cnt: usize,
+    // To avoid de-allocations on the critical path, store cleared number.
+    skip_cleared: usize,
+}
+
+type LOGS = Vec<CachePadded<Mutex<BufferedLog>>>;
+
+// .1 (AtomicUsize type) tracks the total number of speculative critical errors.
+static BUFFERED_LOGS: Lazy<ArcSwap<(LOGS, AtomicUsize)>> =
+    Lazy::new(|| ArcSwap::new(Arc::new((LOGS::new(), AtomicUsize::new(0)))));
```

Comment:
Why keep a total error counter instead of computing the total error number from the counters inside each txn log?

Reply:
to make flush efficient - it returns the number but actual flushing happens asynchronously by the spawned task.